### DISCUSSION TITLE
issue: custom validadtion gets overwritten by zod resolver

### DISCUSSION THREAD

**Discussion by dorbn4** [react-hook-form/react-hook-form#10153] [Q&A]
### Version Number

7.43.7

### Codesandbox/Expo snack

https://codesandbox.io/s/aged-architecture-z8dtdr?file=/src/App.tsx

### Steps to reproduce

See codesandbox :)

Zod and react hook form are amazing libraries. 

**Context:** If you use a resolver in your RHF all of your custom validation (inside the `controller` and `register`) is overwritten and not even applied. 

**Use case:** Numeric `min`/`max` input for e.g. min/max price/amount etc. 

References:
- [Ref 1](https://github.com/react-hook-form/react-hook-form/discussions/7512), 
- [Ref 2](https://react-hook-form.com/api/useform/#resolver:~:text=A%20resolver%20can%20not%20be%20used%20with%20the%20built%2Din%20validators%20(e.g.%3A%20required%2C%20min%2C%20etc.))

### Expected behaviour

My custom validation should overwrite (or extend) the zod resolver.

IF this is expected THEN we should hide the validation method from the `controller`/`register` methods and make it clearer in the documentation. 

BUT I really think that this is a bug and should be resolved (would also be up for a PR but want to first understand the reasoning if there is any)

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
Please refer to the doc below: 

https://react-hook-form.com/api/useform/

> A resolver can not be used with the built-in validators (e.g.: required, min, etc.)

**Reply by dorbn4** [reply]
I know, I added this exact link in my ticket. Could you eloberate as to to why this is the implementation, or how to fix this?

As far as I can see, the only option would be to remove the resolver completely

**Reply by leapful** [reply]
@dorbn4 

The built-in validation rules of react-hook-form are aligned with common web API validation as it's applied on field-level while schema validation library like zod is built to centralize all validation rules as a single source of truth.

Mixing built-in validation and schema validation only makes the implementation of validation handling more complex and leads to more confusion on debugging and solving problems. For example, your use case expects to overwrite or expand zod resolver, but the other developers might want the opposite.

If you validation use case is simple and purely related to common web API validation, just stick with it, otherwise, if it's very complex and requires more customization or transformation, please use schema validation library

**Reply by marianban** [reply]
It would be nice to have the default validation resolver function as an input parameter of the resolver function so the build in validators can still be manually triggered.

**Reply by camin-mccluskey** [reply]
Is there any way to achieve the same outcome without removing the resolver? I've tried using `setError` but run into a similar issue whereby the `isValid` value is still true due to the zod validation.

**Reply by janvorwerk** [reply]
Yes, I am wondering how we could keep a zod schema, yet add another custom validation rule on top.

For instance: a form letting the user change it's email address: I just want to validate that the typed address is not the current one. This does not belong to the data schema... it's just elsewhere.

Indeed, having the validate() function receive another chained validator (as suggested by @marianban )would be perfect: it would let the dev choose exactly when and if the other validators would be called.

**Reply by janvorwerk** [reply]
FWIW, I found a workaround with a watch:
```
  useEffect(() => {
    const subscription = form.watch((values, { type }) => {
      if (values.email === user.email) {
        form.setError("email", { message: "This is already your current address" });
      } else {
        form.clearErrors("email");
        void form.trigger("email");
      }
    });
    return () => subscription.unsubscribe();
  }, [form.watch]);
```
I wish I could only validate on blur, so that the field is not in error after the 1st key stroke... Something like:

    if (type === "blur") {

But the `type` is always `"change"`: the watch callback does not get called on blur events. I had to add the onBlur handler myself and deal with that is a react state hook.

**Comment by Mcoder995** [comment]
Here is my solution:

```
function Form() {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<FormType>({
    resolver: zodResolver(
      zodDemoMinMaxSchema.superRefine((data, ctx) => {
        Object.keys(errors).forEach((error) => {
          ctx.addIssue({
            path: [error],
            message: errors[error as keyof FormType]?.message,
            code: "custom",
          });
        });
      })
    ),
    defaultValues: {
      priceRange: {
        min: 0,
        max: 0,
      },
    },
  });
```

**Reply by simonrey1** [reply]
Thanks @Mcoder995 ! I have changed a bit for my use

```ts
    const form = useForm({
        resolver: zodResolver(
            formSchema.superRefine((data, ctx) => {
                const checkNestedErrors = (errors: FieldErrors, path: string[] = []) => {
                    Object.entries(errors).forEach(([key, error]) => {
                        if (error?.type === 'custom') {
                            console.log(error);
                            ctx.addIssue({
                                path: [...path, key],
                                message: error?.message?.toString(),
                                code: 'custom',
                            });
                        }
                        if (typeof error === 'object' && error !== null && !error.type) {
                            checkNestedErrors(error as FieldErrors, [...path, key]);
                        }
                    });
                };

                checkNestedErrors(form.formState.errors);
            })
        ),
        values,
    });
```

**Reply by Amansaxena001** [reply]
This is one case but consider multiple custom validations that depends upon business logic we cant write all of them here. We would need a scalable mech where we can easily write custom validations that runs in sync with zod resolver
